src/watch.rs

```diff
@@ -0,0 +1,77 @@
+use crate::{cmd, util::Fallible};
+
+use clap::ArgMatches;
+use notify::Watcher;
+use std::{
+    collections::HashSet, env, path::PathBuf, sync::mpsc::channel, thread::sleep, time::Duration,
+};
+
+pub fn watch(
+    args: ArgMatches,
+    watch_sources: HashSet<PathBuf>,
+    watch_code: HashSet<PathBuf>,
+) -> Fallible {
+    trigger_on_change(&watch_sources, &watch_code, |paths| {
+        let is_forward = paths
+            .first()
+            .map(|p| watch_sources.contains(p))
+            .unwrap_or(true);
+
+        println!(
+            "{} changed. Re-building...",
+            if is_forward { "Sources" } else { "Code" }
+        );
+
+        let curr_dir = env::current_dir()?;
+        let (_config, _watch_forward, _watch_reverse) =
+            cmd::run_with_args(&args, Some(!is_forward))?;
+        env::set_current_dir(&curr_dir)?;
+
+        Ok(())
+    })?;
+
+    Ok(())
+}
+
+/// Calls the closure when a book source file is changed, blocking indefinitely.
+pub fn trigger_on_change<F>(
+    watch_sources: &HashSet<PathBuf>,
+    watch_code: &HashSet<PathBuf>,
+    closure: F,
+) -> Fallible
+where
+    F: Fn(Vec<PathBuf>) -> Fallible,
+{
+    use notify::DebouncedEvent::*;
+    use notify::RecursiveMode::*;
+
+    let (tx, rx) = channel();
+    let mut source_watcher = notify::watcher(tx.clone(), Duration::from_secs(1))?;
+    let mut code_watcher = notify::watcher(tx, Duration::from_secs(1))?;
+
+    for path in watch_sources {
+        source_watcher.watch(path, NonRecursive)?;
+    }
+    for path in watch_code {
+        code_watcher.watch(path, NonRecursive)?;
+    }
+
+    loop {
+        let first_event = rx.recv().unwrap();
+        sleep(Duration::from_millis(50));
```

Comment:
This does not look right. I think you want to coalesce all events within 50ms into a single change? How about
```rust
rx.recv().unwrap();
let deadline = Instant::now() + Duration::from_millis(50);

loop {
  let timeout = match deadline.checked_sub(Instant::now()) {
    Some(timeout) => timeout,
    None => break,
  };
  
  rx.recv_timeout(timeout);
}

tx_changes.send(ChangeType::Sources).unwrap();
```
which could be done directly in the post-processing threads suggested above.